### PR TITLE
[#IC-146] Delete CGN subscription

### PR DESCRIPTION
**Pull request by iwoak**

### CODE REVIEW

**Comment by iwoak** [comment]
@AleDore finally I wrote missing unit test by commit [0c74d702cfe3dd6a2282c66576ea75779a6f1486](https://github.com/pagopa/io-functions-cgn/pull/50/commits/0c74d702cfe3dd6a2282c66576ea75779a6f1486)

**Comment by AleDore** [comment]
@iwoak It seems lint fails :)

**Comment by iwoak** [comment]
> @iwoak It seems lint fails :)

Done :( @AleDore

File: DeleteCardActivation/__test__/handler.test.ts [hunk 1]
```diff
@@ -0,0 +1,255 @@
...
+    eycaFindLastVersionByModelIdMock.mockImplementationOnce(() =>
+      taskEither.of(some(aEycaUserCardRevoked))
+    );
+    const deleteCardActivationHandler = DeleteCardActivationHandler(
+      (userEycaModelMock as unknown) as UserEycaCardModel,
+      (userCgnModelMock as unknown) as UserCgnModel
+    );
+    const response = await deleteCardActivationHandler(context, aFiscalCode);
+    expect(response.kind).toBe("IResponseErrorConflict");
+  });
```
[the remark this answers is not in the document]
Comment (AleDore) [reply DeleteCardActivation/__test__/handler.test.ts:168]: In theory this scenario should be possible, but I have no evidence if EYCA provides revocation for its cards.

**Comment by AleDore** [comment]
@iwoak : can you please make a review after fp-ts2 porting?

File: openapi/eyca.yaml [hunk 2]
```diff
@@ -158,7 +158,53 @@ paths:
           description: Service unavailable.
           schema:
             $ref: "#/definitions/CommonEycaResponse"
-  
+  "/delete_card":
```
Comment (michaeldisaro) [inline openapi/eyca.yaml:161]: Fix the base path above this. It should be something like /api/v1/eyca not just /api.

File: openapi/index.yaml [hunk 3]
```diff
@@ -93,6 +94,7 @@ paths:
         Get the EYCA status details by the provided fiscal code. 
         In case of success the response could be one of:
           - CardPending
```
Comment (michaeldisaro) [inline openapi/index.yaml:96]: Isn't the case to refactor to EycaCardPending?

File: DeleteCgnOrchestrator/handler.ts [hunk 4]
```diff
@@ -0,0 +1,299 @@
...
+      if (decodedDeleteCgnResult.kind !== "SUCCESS") {
+        trackExAndThrowWithError(
+          new Error("Cannot delete CGN"),
+          "cgn.delete.exception.failure.cgnActivityOutput"
+        );
+      } else {
+        cgnDataToBackup = decodedDeleteCgnResult.cards;
+
+        // Backup all data for legal issue
+        const legalDataToBackup: DeleteLegalDataBackupActivityInput = {
```
Comment (michaeldisaro) [inline DeleteCgnOrchestrator/handler.ts:200]: What if eyca is deleted and an error is throw by cgn deletion? Will eyca data be recoverable?

**Review by michaeldisaro — APPROVED** [review]
Apart minor changes it LFTM (Looks Feasible To Me).

File: openapi/index.yaml (hunk 3 above)
Comment (AleDore) [reply openapi/index.yaml:96]: Nope, CardPending is a status inherited from CommonCard

File: openapi/eyca.yaml (hunk 2 above)
Comment (AleDore) [reply openapi/eyca.yaml:161]: These are 3rd parties apis so we must let them this way :)

File: DeleteCgnOrchestrator/__tests__/handler.test.ts [hunk 5]
```diff
@@ -0,0 +1,278 @@
...
+    // 4 CreateTimer
+    createTimer: jest.fn().mockReturnValue("CreateTimer")
+  }
+};
+
+describe("DeleteCgnOrchestrator", () => {
+  beforeEach(() => {
+    jest.clearAllMocks();
+  });
+  it("should delete only cgn data while eyca is not present", async () => {
```
Comment (michaeldisaro) [inline DeleteCgnOrchestrator/__tests__/handler.test.ts:86]:
```suggestion
  it("should delete only cgn data when eyca is not present", async () => {
```

**Review by michaeldisaro — APPROVED** [review]

**Pull request merged**